seeker/log/journald.go

```diff
@@ -0,0 +1,126 @@
+package log
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/hcdiag/seeker"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+var _ seeker.Runner = Journald{}
+
+// JournaldTimeLayout custom go time layouts must match the reference time Jan 2 15:04:05 2006 MST
+const JournaldTimeLayout = "2006-01-02 15:04:05"
+
+type Journald struct {
+	service string
+	destDir string
+	since   time.Time
+	until   time.Time
+}
+
+// NewJournald sets the defaults for the journald runner and returns a seeker
+func NewJournald(service, destDir string, since, until time.Time) *seeker.Seeker {
+	return &seeker.Seeker{
+		Identifier: "journald",
+		Runner: Journald{
+			service: service,
+			destDir: destDir,
+			since:   since,
+			until:   until,
+		},
+	}
+}
+
+// Run attempts to pull logs from journald via shell command, e.g.:
+// journalctl -x -u {name} --since '3 days ago' --no-pager > {destDir}/journald-{name}.log
+func (j Journald) Run() (interface{}, seeker.Status, error) {
+	// Check if systemd has a unit with the provided name, return a nil pointer
+	cmd := fmt.Sprintf("systemctl is-enabled %s", j.service) // TODO(gulducat): another command?
+	out, err := seeker.NewCommander(cmd, "string").Run()
+	if err != nil {
+		hclog.L().Debug("skipping journald", "service", j.service, "output", out, "error", err)
+		return nil, seeker.Fail, JournaldServiceNotEnabled{
+			service: j.service,
+			command: cmd,
+			result:  fmt.Sprintf("%s", out),
+			err:     err,
+		}
+	}
+
+	// check if user is able to read messages
+	cmd = fmt.Sprintf("journalctl -n0 -u %s 2>&1 | grep -A10 'not seeing messages from other users'", j.service)
+	out, err = seeker.NewSheller(cmd).Run()
+	if err != nil {
```

Comment:
I'm a little confused on this bit. If I'm reading this right, we'll return when `err != nil` (lines 56-58), and we'll also return when `err == nil` (lines 60-67). I think anything beyond here would be unreachable.

In the original file, it looks like the expectation was that `err` should NOT be nil, so I suspect perhaps lines 56-58 can just be removed.

Reply:
hmmm - yeah that's a bug. fixing in this revision, thanks !